internal/server/auth/method/oauth/server.go

```diff
@@ -0,0 +1,149 @@
+	store  storageauth.Store
+	config config.AuthenticationConfig
+
+	auth.UnimplementedAuthenticationMethodOAuthServiceServer
+}
+
+// NewServer constructs a Server.
+func NewServer(
+	logger *zap.Logger,
+	store storageauth.Store,
+	config config.AuthenticationConfig,
+) *Server {
+	return &Server{
+		logger: logger,
+		store:  store,
+		config: config,
+	}
+}
+
+// RegisterGRPC registers the server as an Server on the provided grpc server.
+func (s *Server) RegisterGRPC(server *grpc.Server) {
+	auth.RegisterAuthenticationMethodOAuthServiceServer(server, s)
+}
+
+func callbackURL(host, oauthHost string) string {
+	// strip trailing slash from host
+	host = strings.TrimSuffix(host, "/")
+	return host + "/auth/v1/method/oauth/" + oauthHost + "/callback"
+}
+
+// AuthorizeURL will return a URL for the client to redirect to for completion of the OAuth flow with GitHub.
+func (s *Server) AuthorizeURL(ctx context.Context, a *auth.OAuthAuthorizeRequest) (*auth.AuthorizeURLResponse, error) {
+	oauthConfig := s.config.Methods.OAuth.Method.Hosts[a.Host]
+
+	authorizeBaseURL := hostToAuthorizeBaseURL[a.Host]
+
+	u, err := url.Parse(authorizeBaseURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build OAuth authorize url.
+	q := u.Query()
+
+	q.Set("client_id", oauthConfig.ClientId)
+	q.Set("scope", strings.Join(oauthConfig.Scopes, ":"))
+	q.Set("redirect_uri", callbackURL(oauthConfig.RedirectAddress, a.Host))
+
+	u.RawQuery = q.Encode()
+
+	return &auth.AuthorizeURLResponse{
+		AuthorizeUrl: u.String(),
+	}, nil
+}
+
+// OAuthCallback is the OAuth callback method for OAuth authentication. It will take in a SessionCode
+// which is the OAuth grant passed in by the OAuth service, and exchange the grant with an Authentication
+// that includes the access token.
+func (s *Server) OAuthCallback(ctx context.Context, oauth *auth.OAuthCallbackRequest) (*auth.OAuthCallbackResponse, error) {
+	oauthConfig := s.config.Methods.OAuth.Method.Hosts[oauth.Host]
+
+	c := &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	accessTokenURL := hostToAccessTokenBaseURL[oauth.Host]
+
+	req, err := http.NewRequestWithContext(ctx, "POST", accessTokenURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	q := req.URL.Query()
+	q.Set("client_id", oauthConfig.ClientId)
+	q.Set("client_secret", oauthConfig.ClientSecret)
+	q.Set("code", oauth.Code)
+	e := q.Encode()
+	req.URL.RawQuery = e
+
+	// We have to accept JSON from the GitHub server when requesting the access token.
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var oauthAccessTokenResponse struct {
+		AccessToken string   `json:"access_token,omitempty"`
+		Scopes      []string `json:"scopes,omitempty"`
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&oauthAccessTokenResponse); err != nil {
+		return nil, err
+	}
+
+	metadata := map[string]string{
+		storageMetadataGithubAccessToken: oauthAccessTokenResponse.AccessToken,
```

Comment:
I'm not sure we're really going to want to store this.
Particularly in this way, which will leak it through the API.
Instead, we should just use it to get identity information and just put the pieces we get from Github in the metadata (profile information etc.). That is what we do in the OIDC method.

Reply:
@GeorgeMac Okay was thinking that too. I did it like this because I was wondering if we wanted to keep that access token around for making authorized requests to GH

Reply:
I think, once we have used it to finish the auth flow and get identity information we can discard it.
Somewhere down the line we might want to re-explore storing this, refresh tokens and so on.
But if we do, we would likely store it somewhere not API accessible.